crates/biome_js_analyze/src/lint/nursery/use_at_index.rs

```diff
@@ -0,0 +1,927 @@
+}
+
+/// The method to retrieve values from `.slice()`
+#[derive(Clone)]
+pub enum SliceExtractType {
+    Pop,
+    Shift,
+    ZeroMember,
+}
+
+/// The number of arguments for `.slice()`
+#[derive(Clone)]
+pub enum SliceArgType {
+    OneArg,
+    TwoArg,
+}
+
+/// Type of Code to Fix
+#[derive(Clone)]
+pub enum ErrorType {
+    Index {
+        is_negative: bool,
+    },
+    StringCharAt {
+        is_negative: bool,
+    },
+    Slice {
+        arg_type: SliceArgType,
+        extract_type: SliceExtractType,
+    },
+    GetLastFunction,
+}
+
+impl ErrorType {
+    /// Return the error message corresponding to the ErrorType.
+    fn get_error_message(self: &ErrorType) -> MarkupBuf {
+        match self {
+            ErrorType::Index { is_negative } | ErrorType::StringCharAt { is_negative } => {
+                let (method, old_method) = if *is_negative {
+                    ("X.at(-Y)", if matches!(self, ErrorType::StringCharAt { .. }) { "X.charAt(X.length - Y)" } else { "X[X.length - Y]" })
+                } else {
+                    ("X.at(Y)", if matches!(self, ErrorType::StringCharAt { .. }) { "X.charAt(Y)" } else { "X[Y]" })
+                };
+                markup! { "Prefer "<Emphasis>{method}</Emphasis>" over "<Emphasis>{old_method}</Emphasis>"." }.to_owned()
+            },
+            ErrorType::Slice { arg_type, extract_type } => {
+                let extract_string = match extract_type {
+                    SliceExtractType::Pop => ".pop()",
+                    SliceExtractType::Shift => ".shift()",
+                    SliceExtractType::ZeroMember => "[0]",
+                };
+                let (method, old_method) = match (arg_type, extract_type) {
+                    (SliceArgType::OneArg, SliceExtractType::Pop) => ("X.at(-1)", format!("X.slice(-a){}", extract_string)),
+                    (SliceArgType::TwoArg, SliceExtractType::Pop) => ("X.at(Y - 1)", format!("X.slice(a, Y){}", extract_string)),
+                    _ => ("X.at(Y)", format!("X.slice({}){}", if matches!(arg_type, SliceArgType::OneArg) { "Y" } else { "Y, a" }, extract_string)),
+                };
+                markup! { "Prefer "<Emphasis>{method}</Emphasis>" over "<Emphasis>{old_method}</Emphasis>"." }.to_owned()
+            },
+            ErrorType::GetLastFunction => {
+                markup! { "Prefer "<Emphasis>"X.at(-1)"</Emphasis>" over "<Emphasis>"_.last(X)"</Emphasis>"." }.to_owned()
+            }
+        }
+    }
+}
+
+declare_node_union! {
+    pub AnyJsArrayAccess = JsComputedMemberExpression | JsCallExpression
+}
+
+pub struct UseAtIndexState {
+    at_number_exp: AnyJsExpression,
+    error_type: ErrorType,
+    object: AnyJsExpression,
+}
+
+impl UseAtIndexState {
+    fn new(at_number_exp: AnyJsExpression, error_type: ErrorType, object: AnyJsExpression) -> Self {
+        Self {
+            at_number_exp,
+            error_type,
+            object,
+        }
+    }
+}
```

Comment:
Since the constructor set all fields, I could remove it and directly initialize the struct.
```suggestion
```